test/src/unit-capi-array_schema.cc

```diff
@@ -2359,6 +2359,222 @@ TEST_CASE_METHOD(
+    rc = tiledb_query_set_data_buffer(
+        ctx_, query, "d1", dim_buffer, &dim_buffer_size);
+    CHECK(rc == TILEDB_OK);
+  }
+  rc = tiledb_query_submit(ctx_, query);
+  CHECK(rc == TILEDB_OK);
+  rc = tiledb_query_finalize(ctx_, query);
+  CHECK(rc == TILEDB_OK);
+  rc = tiledb_array_close(ctx_, array);
+  CHECK(rc == TILEDB_OK);
+  tiledb_array_free(&array);
+  tiledb_query_free(&query);
+
+  // Clean up
+  tiledb_attribute_free(&attr1);
+  tiledb_attribute_free(&attr2);
+  tiledb_dimension_free(&d1);
+  tiledb_domain_free(&domain);
+  tiledb_array_schema_free(&array_schema);
+
+  // Create an array schema evolution
+  tiledb_array_schema_evolution_t* array_schema_evolution;
+  rc = tiledb_array_schema_evolution_alloc(ctx_, &array_schema_evolution);
+  REQUIRE(rc == TILEDB_OK);
+
+  // Remove attribute a1
+  rc = tiledb_array_schema_evolution_drop_attribute(
+      ctx_, array_schema_evolution, "a1");
+  REQUIRE(rc == TILEDB_OK);
+
+  // Set timestamp to avoid race condition
+  uint64_t now = tiledb_timestamp_now_ms();
+  now = now + 1;
+  rc = tiledb_array_schema_evolution_set_timestamp_range(
+      ctx_, array_schema_evolution, now, now);
+
+  // Evolve schema
+  rc = tiledb_array_evolve_wrapper(
+      ctx_, array_name.c_str(), array_schema_evolution);
+  REQUIRE(rc == TILEDB_OK);
+  tiledb_array_schema_evolution_free(&array_schema_evolution);
+
+  // Add back attribute a1 as a string
+  rc = tiledb_array_schema_evolution_alloc(ctx_, &array_schema_evolution);
+  REQUIRE(rc == TILEDB_OK);
+  tiledb_attribute_t* attr3;
+  rc = tiledb_attribute_alloc(ctx_, "a1", TILEDB_STRING_ASCII, &attr3);
+  REQUIRE(rc == TILEDB_OK);
+  rc = tiledb_attribute_set_cell_val_num(ctx_, attr3, TILEDB_VAR_NUM);
+  REQUIRE(rc == TILEDB_OK);
+  rc = tiledb_array_schema_evolution_add_attribute(
+      ctx_, array_schema_evolution, attr3);
+  REQUIRE(rc == TILEDB_OK);
+
+  // Evolve schema
+  now = tiledb_timestamp_now_ms();
+  now = now + 1;
+  rc = tiledb_array_schema_evolution_set_timestamp_range(
+      ctx_, array_schema_evolution, now, now);
+  rc = tiledb_array_evolve_wrapper(
+      ctx_, array_name.c_str(), array_schema_evolution);
+  REQUIRE(rc == TILEDB_OK);
+  tiledb_attribute_free(&attr3);
+  tiledb_array_schema_evolution_free(&array_schema_evolution);
+
+  // Open array for reads
+  rc = tiledb_array_alloc(ctx_, array_name.c_str(), &array);
+  REQUIRE(rc == TILEDB_OK);
+  rc = tiledb_array_set_open_timestamp_end(ctx_, array, now + 1);
+  REQUIRE(rc == TILEDB_OK);
+  rc = tiledb_array_open(ctx_, array, TILEDB_READ);
+  REQUIRE(rc == TILEDB_OK);
+
+  // Read the array
+  int64_t subarray_read[] = {1, 10};
+  tiledb_subarray_t* sub;
+  rc = tiledb_subarray_alloc(ctx_, array, &sub);
+  CHECK(rc == TILEDB_OK);
+  rc = tiledb_subarray_set_subarray(ctx_, sub, subarray_read);
+  CHECK(rc == TILEDB_OK);
+  rc = tiledb_query_alloc(ctx_, array, TILEDB_READ, &query);
+  CHECK(rc == TILEDB_OK);
+  rc = tiledb_query_set_layout(ctx_, query, TILEDB_ROW_MAJOR);
```

Comment:
For sparse, I think we want to read both unordered and global order, so that we hit both of our sparse readers.